pkg/controller/pod/pod_controller.go

```diff
@@ -72,90 +71,77 @@ var _ reconcile.Reconciler = &ReconcilePod{}
+	if r.sharedInfo.AntreaAgentDaemonSetSpec != nil {
+		if request.Name == r.sharedInfo.AntreaAgentDaemonSetSpec.GetName() && request.Namespace == r.sharedInfo.AntreaAgentDaemonSetSpec.GetNamespace() {
+			return true
+		}
 	}
-
-	// Check if this Pod already exists
-	found := &corev1.Pod{}
-	err = r.client.Get(context.TODO(), types.NamespacedName{Name: pod.Name, Namespace: pod.Namespace}, found)
-	if err != nil && errors.IsNotFound(err) {
-		reqLogger.Info("Creating a new Pod", "Pod.Namespace", pod.Namespace, "Pod.Name", pod.Name)
-		err = r.client.Create(context.TODO(), pod)
-		if err != nil {
-			return reconcile.Result{}, err
+	if r.sharedInfo.AntreaControllerDeploymentSpec != nil {
+		if request.Name == r.sharedInfo.AntreaControllerDeploymentSpec.GetName() && request.Namespace == r.sharedInfo.AntreaControllerDeploymentSpec.GetNamespace() {
+			return true
 		}
-
-		// Pod created successfully - don't requeue
-		return reconcile.Result{}, nil
-	} else if err != nil {
-		return reconcile.Result{}, err
 	}
-
-	// Pod already exists - don't requeue
-	reqLogger.Info("Skip reconcile: Pod already exists", "Pod.Namespace", found.Namespace, "Pod.Name", found.Name)
-	return reconcile.Result{}, nil
+	return false
 }
 
-// newPodForCR returns a busybox pod with the same name/namespace as the cr
-func newPodForCR(cr *corev1.Pod) *corev1.Pod {
-	labels := map[string]string{
-		"app": cr.Name,
+func (r *ReconcilePod) recreateResourceIfNotExist(request *reconcile.Request) error {
+	r.sharedInfo.Lock()
+	defer r.sharedInfo.Unlock()
+	var curObject runtime.Object
+	var objectSpec *uns.Unstructured
+	if request.Name == r.sharedInfo.AntreaAgentDaemonSetSpec.GetName() && request.Namespace == r.sharedInfo.AntreaAgentDaemonSetSpec.GetNamespace() {
+		curObject = &appsv1.DaemonSet{}
+		objectSpec = r.sharedInfo.AntreaAgentDaemonSetSpec.DeepCopy()
+	} else {
+		curObject = &appsv1.Deployment{}
+		objectSpec = r.sharedInfo.AntreaControllerDeploymentSpec.DeepCopy()
 	}
-	return &corev1.Pod{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      cr.Name + "-pod",
-			Namespace: cr.Namespace,
-			Labels:    labels,
-		},
-		Spec: corev1.PodSpec{
-			Containers: []corev1.Container{
-				{
-					Name:    "busybox",
-					Image:   "busybox",
-					Command: []string{"sleep", "3600"},
-				},
-			},
-		},
+	err := r.client.Get(context.TODO(), request.NamespacedName, curObject)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info(fmt.Sprintf("K8s resource - '%s' dose not exist", request.Name))
+		} else {
+			log.Error(err, fmt.Sprintf("Could not retrieve K8s resource - '%s'", request.Name))
+			r.status.SetDegraded(statusmanager.OperatorConfig, "ApplyObjectsError", fmt.Sprintf("Failed to apply objects: %v", err))
+			return err
+		}
+	} else {
+		log.Info(fmt.Sprintf("K8s resource - '%s' already exists", request.Name))
+		return nil
 	}
+	if err = apply.ApplyObject(context.TODO(), r.client, objectSpec); err != nil {
```

Comment:
Set owner here as well

Reply:
Hi @dantingl,
I moved the set owner operation in function `updateStatusManagerAndSharedInfo`. In the function, the obj will be set owner and be shared with PodController through:
- r.sharedInfo.AntreaAgentDaemonSetSpec
- r.sharedInfo.AntreaControllerDeploymentSpec

Reply:
Got it, thanks